### PR TITLE
GEODE-8702: Add StringPrefixPartitionResolver

### PR DESCRIPTION
**Pull request by gaussianrecurrence**
- Added StringPrefixPartitionResolver as it is included on the Java API
 - UTs were added to verify the PR functionality.

### CODE REVIEW

**Comment by gaussianrecurrence** [comment]
Maybe @moleske @pivotal-jbarrett @mreddington @pdxcodemonkey you could throw some comments?

All test passed as can be seen here: https://github.com/gaussianrecurrence/geode-native/runs/1391210631 (Thanks @moleske for GitHub actions pipeline)

Thanks!

**Review by mreddington — APPROVED** [review]

**Review by echobravopapa — COMMENTED** [review]
is this dependent on GEODE-8700?

**Comment by gaussianrecurrence** [comment]
> is this dependent on GEODE-8700?

Not really... one could use this PartitionResolver by instantiating the cache programatically or by using a shared library.

File: cppcache/include/geode/StringPrefixPartitionResolver.hpp [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+ */
+class APACHE_GEODE_EXPORT StringPrefixPartitionResolver
+    : public PartitionResolver {
+ public:
+  StringPrefixPartitionResolver();
+  explicit StringPrefixPartitionResolver(std::string delimiter);
+
+  StringPrefixPartitionResolver(const StringPrefixPartitionResolver&) = delete;
+
+  /**
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/StringPrefixPartitionResolver.hpp:80]: Delete redundant comments.

File: cppcache/include/geode/StringPrefixPartitionResolver.hpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace apache {
+namespace geode {
+namespace client {
+
+class CacheableKey;
+class EntryEvent;
+
+/**
+ * Implement the <code>PartitionResolver</code> interface to enable custom
```
Comment (pivotal-jbarrett) [inline cppcache/include/geode/StringPrefixPartitionResolver.hpp:33]: It looks like you copied the documentation without updating it for this specific implementation. Please change documentation to be consistent with the implementation. Should probably mention the Java version of this resolver.

**Review by pivotal-jbarrett — CHANGES_REQUESTED** [review]

File: cppcache/include/geode/StringPrefixPartitionResolver.hpp (hunk 2 above)
Comment (gaussianrecurrence) [reply cppcache/include/geode/StringPrefixPartitionResolver.hpp:33]: Ups, sorry. Thanks for pointing it out :)

**Review by pivotal-jbarrett — APPROVED** [review]

**Comment by pdxcodemonkey** [comment]
@gaussianrecurrence This breaks the RHEL7 build, so the change has been reverted in `develop`, sorry.  Please open a new PR when resolved.  Here's the relevant output from the RHEL7 build:

In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc: In instantiation of ‘static _CharT* std::basic_string<_CharT, _Traits, _Alloc>::_S_construct(std::basic_string<_CharT, _Traits, _Alloc>::size_type, _CharT, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = apache::geode::client::size_tracking_allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:665:31:   required from ‘std::basic_string<_CharT, _Traits, _Alloc>::basic_string(std::basic_string<_CharT, _Traits, _Alloc>::size_type, _CharT, const _Alloc&) [with _CharT = char; _Traits = std::char_traits<char>; _Alloc = apache::geode::client::size_tracking_allocator<char>; std::basic_string<_CharT, _Traits, _Alloc>::size_type = long unsigned int]’
/home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:243:35:   required from here
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: error: no match for ‘operator==’ (operand types are ‘const apache::geode::client::size_tracking_allocator<char>’ and ‘apache::geode::client::size_tracking_allocator<char>’)
       if (__n == 0 && __a == _Alloc())
                           ^
In file included from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:33:0:
/home/geode/nativeclient/cppcache/shared/../src/util/size_tracking_allocator.hpp:92:8: note: candidate: bool apache::geode::client::size_tracking_allocator<_Tp>::operator==(const apache::geode::client::size_tracking_allocator<_Tp>&) [with _Tp = char] <near match>
   bool operator==(size_tracking_allocator const &) { return true; }
        ^
/home/geode/nativeclient/cppcache/shared/../src/util/size_tracking_allocator.hpp:92:8: note:   passing ‘const apache::geode::client::size_tracking_allocator<char>*’ as ‘this’ argument discards qualifiers
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/list:63:0,
                 from /home/geode/nativeclient/cppcache/shared/../src/PdxType.hpp:23,
                 from /home/geode/nativeclient/cppcache/shared/../src/PdxTypeRegistry.hpp:33,
                 from /home/geode/nativeclient/cppcache/shared/../src/CacheImpl.hpp:37,
                 from /home/geode/nativeclient/cppcache/shared/../src/DataOutputInternal.hpp:26,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:30:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:1807:5: note: candidate: template<class _Tp, class _Alloc> bool std::operator==(const std::list<_Tp, _Alloc>&, const std::list<_Tp, _Alloc>&)
     operator==(const list<_Tp, _Alloc>& __x, const list<_Tp, _Alloc>& __y)
     ^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:1807:5: note:   template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note:   ‘const apache::geode::client::size_tracking_allocator<char>’ is not derived from ‘const std::list<_Tp, _Alloc>’
       if (__n == 0 && __a == _Alloc())
                           ^
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/list:63:0,
                 from /home/geode/nativeclient/cppcache/shared/../src/PdxType.hpp:23,
                 from /home/geode/nativeclient/cppcache/shared/../src/PdxTypeRegistry.hpp:33,
                 from /home/geode/nativeclient/cppcache/shared/../src/CacheImpl.hpp:37,
                 from /home/geode/nativeclient/cppcache/shared/../src/DataOutputInternal.hpp:26,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:30:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:285:5: note: candidate: template<class _Val> bool std::operator==(const std::_List_iterator<_Tp>&, const std::_List_const_iterator<_Val>&)
     operator==(const _List_iterator<_Val>& __x,
     ^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_list.h:285:5: note:   template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note:   ‘const apache::geode::client::size_tracking_allocator<char>’ is not derived from ‘const std::_List_iterator<_Tp>’
       if (__n == 0 && __a == _Alloc())
                           ^
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/unordered_set:48:0,
                 from /home/geode/nativeclient/cppcache/include/geode/CacheableKey.hpp:25,
                 from /home/geode/nativeclient/cppcache/include/geode/CacheableBuiltins.hpp:30,
                 from /home/geode/nativeclient/cppcache/include/geode/PdxInstanceFactory.hpp:25,
                 from /home/geode/nativeclient/cppcache/include/geode/RegionService.hpp:26,
                 from /home/geode/nativeclient/cppcache/include/geode/GeodeCache.hpp:23,
                 from /home/geode/nativeclient/cppcache/include/geode/Cache.hpp:27,
                 from /home/geode/nativeclient/cppcache/include/geode/CacheFactory.hpp:25,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:25:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1421:5: note: candidate: template<class _Value, class _Hash, class _Pred, class _Alloc> bool std::operator==(const std::unordered_multiset<_Value, _Hash, _Pred, _Alloc>&, const std::unordered_multiset<_Value, _Hash, _Pred, _Alloc>&)
     operator==(const unordered_multiset<_Value, _Hash, _Pred, _Alloc>& __x,
     ^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1421:5: note:   template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note:   ‘const apache::geode::client::size_tracking_allocator<char>’ is not derived from ‘const std::unordered_multiset<_Value, _Hash, _Pred, _Alloc>’
       if (__n == 0 && __a == _Alloc())
                           ^
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/unordered_set:48:0,
                 from /home/geode/nativeclient/cppcache/include/geode/CacheableKey.hpp:25,
                 from /home/geode/nativeclient/cppcache/include/geode/CacheableBuiltins.hpp:30,
                 from /home/geode/nativeclient/cppcache/include/geode/PdxInstanceFactory.hpp:25,
                 from /home/geode/nativeclient/cppcache/include/geode/RegionService.hpp:26,
                 from /home/geode/nativeclient/cppcache/include/geode/GeodeCache.hpp:23,
                 from /home/geode/nativeclient/cppcache/include/geode/Cache.hpp:27,
                 from /home/geode/nativeclient/cppcache/include/geode/CacheFactory.hpp:25,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:25:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1409:5: note: candidate: template<class _Value, class _Hash, class _Pred, class _Alloc> bool std::operator==(const std::unordered_set<_Value, _Hash, _Pred, _Alloc>&, const std::unordered_set<_Value, _Hash, _Pred, _Alloc>&)
     operator==(const unordered_set<_Value, _Hash, _Pred, _Alloc>& __x,
     ^
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/unordered_set.h:1409:5: note:   template argument deduction/substitution failed:
In file included from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/string:53:0,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/random:40,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/stl_algo.h:66,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/algorithm:62,
                 from /opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/regex:38,
                 from /home/geode/nativeclient/cppcache/test/CacheableStringTests.cpp:21:
/opt/rh/devtoolset-4/root/usr/include/c++/5.3.1/bits/basic_string.tcc:600:27: note:   ‘const apache::geode::client::size_tracking_allocator<char>’ is not derived from ‘const std::unordered_set<_Value, _Hash, _Pred, _Alloc>’
       if (__n == 0 && __a == _Alloc())
                           ^

**Comment by pdxcodemonkey** [comment]
Accidentally reverted the wrong PR - this one is fine, so I re-merged.

**Pull request merged**